### DISCUSSION TITLE
One Folder, One Program, Multiple Executables

### DISCUSSION THREAD

**Discussion by real-yfprojects** [pyinstaller/pyinstaller#5393] [Help] [answered]
I am developing a set of PyQt5 application that all use the same underlying python module. They therefore have the same dependencies. I would like to bundle them all into one folder having one executable for each Qt app. 
The multipackage bundling seems to create a application and folder for each executable while providing hard coded links to the common dependencies. It is said that this slows down execution.
My idea is that the executables simply execute parts of the bundled application. This should be as fast as a standard bundle.

- Is this possible using PyInstaller? 
- And in that case, how is it working?
- If not, is there a way to achieve a similar result?

**Comment by rokm** [comment, marked as answer]
Sounds like you might want to use a combined `COLLECT` - see https://github.com/pyinstaller/pyinstaller/issues/1358#issuecomment-421430592. I've done something like that for a project consisting of a CLI processing program, a GUI processing program, and a results viewer that shared a large portion of the python code and external dependencies. 

Although now that I now a bit more about internals of PyInstaller, I suspect that this approach results in duplication of .pyc modules within the PYZ archives that are embedded in the executables. But that's the flip side of not having all .pyc modules collected in one executable's PYZ and having others refer to it...

**Reply by real-yfprojects** [reply]
So does each executable contain the application script(s) defined in `Analysis`, but dependencies are only stored in the folder created by `COLLECT`?

<details>
  <summary>a default spec file</summary>

```python3
block_cipher = None

a = Analysis(['app.py'],
             pathex=['/home/.../PyInstaller-Example'],
             binaries=[],
             datas=[],
             hiddenimports=[],
             hookspath=[],
             runtime_hooks=[],
             excludes=[],
             win_no_prefer_redirects=False,
             win_private_assemblies=False,
             cipher=block_cipher,
             noarchive=False)

pyz = PYZ(a.pure, a.zipped_data,
             cipher=block_cipher)

exe = EXE(pyz,
          a.scripts,
          [],
          exclude_binaries=True,
          name='app',
          debug=False,
          bootloader_ignore_signals=False,
          strip=False,
          upx=True,
          console=True )

coll = COLLECT(exe,
               a.binaries,
               a.zipfiles,
               a.datas,
               strip=False,
               upx=True,
               upx_exclude=[],
               name='app')
```
</details>

Maybe it would be helpful to include a proper documentation of the `PYZ`, `EXE` and `COLLECT` classes in the [manual][manual-spec-ops]

[manual-spec-ops]: https://pyinstaller.readthedocs.io/en/latest/spec-files.html#spec-file-operation "Read The Docs PyInstaller"

**Reply by rokm** [reply]
> So does each executable contain the application script(s) defined in `Analysis`, but dependencies are only stored in the folder created by `COLLECT`?

Well, strictly speaking, an executable contains the data that is passed into the `EXE` structure. Which in this case, contains collected scripts from `Analysis` (`a.scripts`), and a `PYZ` archive (which contains entries from `a.pure` and `a.zipped_data`). So the end effect is that all python modules are collected into PYZ and embedded in the executable. To see what this encompasses, you can check `build/app/PYZ-00.toc`. In addition to PYZ archive, the executable also contains the entry-point python script (`app.py` in your example) and some PyInstaller python modules - as listed in build/app/EXE-00.toc.

In my combined COLLECT example, I use a separate `Analysis`, `PYZ`, and `EXE` instance for each executable, and then `COLLECT` that combines them, i.e.:

<details>
<summary>spec file</summary>
<p>

```python
# -*- mode: python ; coding: utf-8 -*-

block_cipher = None

# First executable
analysis_a = Analysis(['application_a.py'],
             pathex=['/home/.../src'],
             binaries=[],
             datas=[],
             hiddenimports=[],
             hookspath=[],
             runtime_hooks=[],
             excludes=[],
             win_no_prefer_redirects=False,
             win_private_assemblies=False,
             cipher=block_cipher,
             noarchive=False)
pyz_a = PYZ(analysis_a.pure, analysis_a.zipped_data,
             cipher=block_cipher)
exe_a = EXE(pyz_a,
          analysis_a.scripts,
          [],
          exclude_binaries=True,
          name='application_a',
          debug=False,
          bootloader_ignore_signals=False,
          strip=False,
          upx=True,
          console=True)

# Second executable
analysis_b = Analysis(['application_b.py'],
             pathex=['/home/.../src'],
             binaries=[],
             datas=[],
             hiddenimports=[],
             hookspath=[],
             runtime_hooks=[],
             excludes=[],
             win_no_prefer_redirects=False,
             win_private_assemblies=False,
             cipher=block_cipher,
             noarchive=False)
pyz_b = PYZ(analysis_b.pure, analysis_b.zipped_data,
             cipher=block_cipher)
exe_b = EXE(pyz_b,
          analysis_b.scripts,
          [],
          exclude_binaries=True,
          name='application_b',
          debug=False,
          bootloader_ignore_signals=False,
          strip=False,
          upx=True,
          console=True)

# Collect
coll = COLLECT(exe_a,
               analysis_a.binaries,
               analysis_a.zipfiles,
               analysis_a.datas,
               exe_b,
               analysis_b.binaries,
               analysis_b.zipfiles,
               analysis_b.datas,
               strip=False,
               upx=True,
               upx_exclude=[],
               name='full_project')
```

</p>
</details>


So each executable ends up with a PYZ that contains python modules that correspond to their own analysis, plus their entry-point script. And the executables are collected into common dist directory, along with their external dependencies (e.g., PyQt5 shared libraries). Or more formally, the `COLLECT` is collecting both executables (`exe_a` and `exe_b`) plus the external data and binary files that came from their `Analysis` instances (`.binaries`,   `.zipfiles`, `.datas`).

**Reply by real-yfprojects** [reply]
That means that the *python modules* *in the standard library* as well as other python modules *the application depends on* are included in the executable. The PyQt shared libraries aren't python-only and they therefore are collected separately.
The binaries in the folder aren't duplicated, it seems, but the python modules/packages are.

What does `Analysis().zipped_data` contain and is it needed in the PYZ archive?

**Reply by rokm** [reply]
> That means that the python modules in the standard library as well as other python modules the application depends on are included in the executable. The PyQt shared libraries aren't python-only and they therefore are collected separately.
The binaries in the folder aren't duplicated, it seems, but the python modules/packages are.

Indeed. But you could avoid python modules/package duplication as well by passing `noarchive=True` to all `Analysis` instances. Then the python modules should be collected into .pyc files in the dist folder instead of executable-embedded PYZ archives.

> What does Analysis().zipped_data contain and is it needed in the PYZ archive?

This one seems related to `egg` files, and will probably be empty if none of your modules/packages are packaged as `eggs`.

**Reply by real-yfprojects** [reply]
> Indeed. But you could avoid python modules/package duplication as well by passing `noarchive=True` to all `Analysis` instances. Then the python modules should be collected into .pyc files in the dist folder instead of executable-embedded PYZ archives.

This works. It gets very messy, but you could create a root folder containing a link to the executables. 
*[Edit]* It seems to be not that easy.

<details>
<summary>While trying that a encountered a different problem.</summary>

The build directory isn't updated *(completely)* when changing the .spec file.
In this Case I changed back to `noarchive=False` but that didn't have any effect. The `PYZ-00.toc` file remained empty.
</details>

**Reply by rokm** [reply]
> The build directory isn't updated _(completely)_ when changing the .spec file.
> In this Case I changed back to `noarchive=False` but that didn't have any effect. The `PYZ-00.toc` file remained empty.

Try running `pyinstaller --clean combined_project.spec`. The `--clean` argument should force re-build of everything even if build directory is non-empty.

**Reply by real-yfprojects** [reply]
> Try running `pyinstaller --clean combined_project.spec`. The `--clean` argument should force re-build of everything even if build directory is non-empty.

So this behaviour is intended?
I worked around it by deleting the corresponding folders.

**Reply by real-yfprojects** [reply]
And do you know a way to resolve the mess created by specifying the no archive mode?
Maybe by excluding unneeded packages/modules of the standard library.

**Reply by rokm** [reply]
> So this behaviour is intended?

Avoiding a rebuild unless the code changes is probably intended, but not rebuilding when only spec changes is likely not. 

At any rate, I always run `pyinstaller` with `--clean` option, just in case...

**Reply by real-yfprojects** [reply]
> Avoiding a rebuild unless the code changes is probably intended, but not rebuilding when only spec changes is likely not.

Should a open an Issue?

**Reply by rokm** [reply]
>And do you know a way to resolve the mess created by specifying the no archive mode?


If you are referring to the fact that all .pyc modules from standard library end up in the same directory as executables, then I'm afraid I don't have a solution for that. For my case, duplication of .pyc modules in PYZ archives of multiple executables was a non-issue (because their size was dwarfed by that of external shared libraries), so I never really had to look beyond that. 

> Maybe by excluding unneeded packages/modules of the standard library.

For what is worth, though, I don't imagine that excluding some of those modules would drastically improve the situation.

**Reply by rokm** [reply]
> Should a open an Issue?

Sure, go ahead. If nothing else, someone else might offer a better explanation for the behavior.

**Reply by bwoodsend** [reply]
I believe `--clean` is not required for changes to the spec. If you change `hiddenimports` then PyInstaller understands that it needs to redo everything. If you change either `datas` or `binaries` then PyInstaller doesn't need to redo any analysis - just add or remove the different files. The only times I've found you have to use `--clean` are:
- Any changes to hooks. That includes adding new hooks.
- Any changes to your environment. i.e. Installing/uninstalling/upgrading any package including PyInstaller. 

The second hits us whenever we suggest to someone *try this development branch of PyInstaller*, we must tell them to use `--clean`.

**Reply by real-yfprojects** [reply]
> > Maybe by excluding unneeded packages/modules of the standard library.
> 
> For what is worth, though, I don't imagine that excluding some of those modules would drastically improve the situation.

I now excluded most of the ones that aren't needed for my example script. The folder size decreased from initial 14.2 MB to 7.8 MB.
Unfortunately,  there still are too many files in the folder to find the executable fast.

**Reply by rokm** [reply]
> I believe `--clean` is not required for changes to the spec. If you change `hiddenimports` then PyInstaller understands that it needs to redo everything. If you change either `datas` or `binaries` then PyInstaller doesn't need to redo any analysis - just add or remove the different files.

Hmm, you're right. Changing `noarchive` might be a special case, then. It is not included in  [Analysis._GUTS](https://github.com/pyinstaller/pyinstaller/blob/cc49966ac98e6db18d9e2422e430086f416b8641/PyInstaller/building/build_main.py#L256-L281), either, so that might be the cause. (Unless this particular change should be caught at some other level?).
